### PR TITLE
[data][docs][train] Training ingest example for read_images

### PR DESCRIPTION
**Pull request by stephanie-wang**
## Why are these changes needed?

Adds an example to Ray docs for training ingest on a single node. Walkthrough for adding preprocessor functions, extracting labels, running with Ray Train.

TODOs:
- Make S3 bucket public so you don't need a Kaggle account to run it

Note that this example is somewhat affected by the following issues but we should probably merge it first:
- #38021
- #38763 (prevents the example from using map)
- #38764

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by stephanie-wang** [comment]
For reference, here is a list of issues that came up during this example:
- [[data] Obscure error message for read_images if image dimensions don't match](https://github.com/ray-project/ray/issues/38050)
- [[data] read_images().take(1) produces scary errors](https://github.com/ray-project/ray/issues/38021)
- [[data] read_images().take(1) is very slow on S3 / pushdown limit() into individual tasks](https://github.com/ray-project/ray/issues/38023)
- Improve label extraction - right now the label extraction is very slow because it unnecessarily reads the image data and it has to perform a full shuffle
- [[data] fn_kwargs does not work with ActorPoolStrategy](https://github.com/ray-project/ray/issues/38092)

File: doc/source/ray-air/examples/training_ingest_object_classification.ipynb [hunk 1]
```diff
@@ -0,0 +1,1232 @@
...
...jDg45h8+3jWtvZKOM6Q1AMZCuQtEiSlOkDn2bjx0CmSY1kGXYGWcDhMgMKBAYedxRsfvpkaEqzJ0StDsT79o2jvbKRHOk2smIG5fb1SE7V1uGzbJQRbRQUmBWPGRZtq2nlLnUU7ETSI+1tcV9GwZhQqgWjUSCI22UiqSUIqivTodkSwYyhsNULYDBRGzSMhVTFIkys9zP5fhwfPhZefh8PX7SD5+2coqIKXGgb6i9Z4ttsjIf7tWXFKYns2U6c+0iEvNO/FGd6ynBCB762tsS2d17eEQb3s8pu9hwUMVY52m0t3tjdeXyuvvgwApRVKDmGUm6+8CCnp6erhKFKiJq0rMhQ1TncsjN4TEV88w9FhchARFKilJbG/fxj8ywAJTx7MPm73rUCCSVZacQ4HoNkgB1kI2Ze9YOx5KARQs5rszgDm/zuuvbx7xhHf2TFmmvz8TotpTBvS/a50Neu6QffV29518kS9+MtDDAr50VwT1Ie7zWVUC0S4UQg8Caw7ozkNmjNRcxs5ou5e6TZVtmMxO7PK9/9PNf/tGXb0rm26/+Q18uT++/f373dcS5+9Z6XBY/3h+KGSN69oht25YMxSTn5661FJugnKaiECQtS622bE2pHgLTUsrWV9FEVwrNrGdYNfFuKhTPBCPGliCrZOoYrsQtRSKyRxhspMUlCVOH6FSm4zRNtVYD+7YtkpLeSD49PU2Hu8vS6+GubT2SIiNnau9qj9OBUGqIqqhEjF3MuAEgbGoqqiIHku7NVFUoNrJAx8pmcAlAoKrs14cA18hfABLCTEiaKMPJLJpAKi2SpWggzJDeVIuBWuuQPYeEkMJEBtCLzYmsZSZpZp6hFoEo+wbZKIP9vS+y90NXbKiWh/WwqJZiqiqGwXYc1msgVCtQ1mVbLg1ArVUVotGX6G1t69r9LIpShhUnjENwSEEffW96hERb0kqaZSIxOhowmfN0kMlK0cmmOt9Pr760T7/E3Wu7+3XWuac0CugqgGYkZYtjOWQtZERwaRtAFWSkCbeIHit0LnbIZGS67LhVKUVMkRgS5VvfMAwqrlxGHSareeVzD5IHkp4uArOhoqQpxIxqTPUEEe/evSdJBsTVxH34aJnp/ObV5+/evbMsQrgYx89+nRr2XWS54VmqRbSM9SBL3uIhhnpqfKkWKQQgORJ0doIJ8lB3DCt32z8dP+suIYu
```
Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:339]: Maybe we should always recommend using this approach? Ideally, there shouldn't be multiple ways to do things, and I think people struggle to understand `Partitioning`.

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:482]: To simplify the UDF code, should we use map here? The UDF isn't vectorized, so I don't think we'll get a meaningful performance improvement (?)

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:638]: > To simplify the UDF code, should we use map here? The UDF isn't vectorized, so I don't think we'll get a meaningful performance improvement (?)

Same here

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:746]: > To simplify the UDF code, should we use map here? The UDF isn't vectorized, so I don't think we'll get a meaningful performance improvement (?)

Same here

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:1066]: The train dataset is local but the validation dataset is S3. Is that expected?

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:725]: This isn't rendered as code text on GitHub.

File: doc/source/ray-air/examples/training_ingest_object_classification.ipynb [hunk 2]
```diff
@@ -0,0 +1,1232 @@
...
+    "!pip install \"ray[data]\" torchvision awscli"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Let's first download an image dataset to use. If you already have an image dataset on local disk or in S3, you can skip the rest of this section.\n",
+    "Otherwise, follow on to download a dataset from Kaggle.\n",
+    "For this example, we'll use bash to download a [4GB subset of the ImageNet dataset](https://www.kaggle.com/datasets/ifigotin/imagenetmini-1000) and (optionally) upload it to an S3 bucket.\n",
```
[the remark this answers is not in the document]
Comment (stephanie-wang) [reply doc/source/ray-air/examples/training_ingest_object_classification.ipynb:83]: Ah yeah good point, I think it's best if we include instructions for both.

File: doc/source/ray-air/examples/training_ingest_object_classification.ipynb (hunk 1 above)
Comment (stephanie-wang) [reply doc/source/ray-air/examples/training_ingest_object_classification.ipynb:339]: I agree with you (ideally I wish we didn't have Partitioning) but wasn't sure about what the recommendation is for users. @amogkam what do you think?

Comment (stephanie-wang) [reply doc/source/ray-air/examples/training_ingest_object_classification.ipynb:482]: Actually I'm going to keep this as is if that's okay. Ran into https://github.com/ray-project/ray/issues/38763 and I think generally we may want to guide users towards map_batches for now for performance reasons.

Comment (stephanie-wang) [reply doc/source/ray-air/examples/training_ingest_object_classification.ipynb:1066]: Added a comment here.

Comment (stephanie-wang) [reply doc/source/ray-air/examples/training_ingest_object_classification.ipynb:725]: Ah yeah this was actually intentional because I don't want that code to run as part of the CI tests. Is there another way to do that?

**Comment by amogkam** [comment]
@stephanie-wang +1 to deduping this example from https://docs.ray.io/en/latest/ray-air/examples/torch_detection.html#fine-tuning-a-torch-object-detection-model.

Would prefer to have 1 canonical object detection training example instead of 2. 

What are the main gaps in that example? Would we be able to just fill those gaps?

File: doc/source/ray-air/examples/training_ingest_object_classification.ipynb [hunk 3]
```diff
@@ -0,0 +1,1232 @@
...
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "\u001b[33mWARNING: Ignoring invalid distribution -otocore (/home/ray/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
```
Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:30]: Nit: Remove these outputs? Don't think these are useful to the reader.

File: doc/source/ray-air/examples/training_ingest_object_classification.ipynb [hunk 4]
```diff
@@ -0,0 +1,1232 @@
...
+    "# ...\n",
+    "# └── val\n",
+    "#     ├── n01440764\n",
+    "#     │   ├── ILSVRC2012_val_00009111.JPEG\n",
+    "#     │   ├── ILSVRC2012_val_00030740.JPEG\n",
+    "# ...\n",
+    "```\n",
+    "\n",
+    "If you want to read from S3, run the following commands and pass the S3 URI `s3://imagenetmini-1000/train` in the following Python code.\n",
+    "Otherwise, you can pass the local path `/tmp/imagenet-mini/train`.\n",
```
Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:112]:
```suggestion
    "If you want to read from S3, run the following commands and pass the S3 URI `s3://imagenetmini-1000/train` in the following Python code.\n",
    "Otherwise, pass the local path `/tmp/imagenet-mini/train`.\n",
```

File: doc/source/ray-air/examples/training_ingest_object_classification.ipynb [hunk 5]
```diff
@@ -0,0 +1,1232 @@
...
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "^C\n",
+      "fatal error: \n"
+     ]
+    }
+   ],
+   "source": [
+    "!aws s3 sync s3://imagenetmini-1000/ /tmp/imagenetmini-1000"
```
Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:152]: I'm confused. Didn't we already download the dataset in the previous step?

File: doc/source/ray-air/examples/training_ingest_object_classification.ipynb [hunk 6]
```diff
@@ -0,0 +1,1232 @@
...
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "^C\n",
+      "fatal error: \n"
```
Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:147]: Nit: Remove?

File: doc/source/ray-air/examples/training_ingest_object_classification.ipynb (hunk 1 above)
Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:230]: Nit: Here and elsewhere, should we use a Jupyter output instead of illustrating the output in a comment? Might be easier to read, and would avoid duplication 
image

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:414]: There was an error when you ran this cell. Re-run the cell?

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:689]: Any reason we can't use `torch.as_tensor` here to avoid copies?

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:695]: Instead of doing `torch.Tensor` + `np.transpose`, we could just add `transforms.ToTensor` to the transform, right?

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:725]: Think this should be a code cell and not markdown cell?

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:813]: Remove "Ray AIR" since it's being rebranded?

Comment (bveeramani) [inline doc/source/ray-air/examples/training_ingest_object_classification.ipynb:830]: Replace with updated Train APIs?

**Comment by angelinalg** [comment]
@stephanie-wang I'm so sorry for not getting to this earlier. I made some copy edits and tried to put it into a commit to make it easier for you, but I must've done something wrong. Here is a PR with my commit: https://github.com/ray-project/ray/pull/40176

The code examples have references to the Ray AIR API which has been move to Ray Train. Are you still planning to merge this PR? If so, let me know and I can work on updating the code examples.
